x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/analytics/process/AnalyticsProcessManager.java

```diff
@@ -41,28 +42,39 @@ public AnalyticsProcessManager(Client client, Environment environment, ThreadPoo
         this.processFactory = Objects.requireNonNull(analyticsProcessFactory);
     }
 
-    public void processData(String jobId, DataFrameDataExtractor dataExtractor) {
+    public void runJob(String jobId, DataFrameDataExtractorFactory dataExtractorFactory) {
         threadPool.generic().execute(() -> {
-            AnalyticsProcess process = createProcess(jobId, dataExtractor);
-            try {
-                writeHeaderRecord(dataExtractor, process);
-                writeDataRows(dataExtractor, process);
-                process.writeEndOfDataMessage();
-                process.flushStream();
+            DataFrameDataExtractor dataExtractor = dataExtractorFactory.newExtractor(false);
+            AnalyticsProcess process = createProcess(jobId, createProcessConfig(dataExtractor));
+            ExecutorService executorService = threadPool.executor(MachineLearning.AUTODETECT_THREAD_POOL_NAME);
+            AnalyticsResultProcessor resultProcessor = new AnalyticsResultProcessor(client, dataExtractorFactory.newExtractor(true));
+            executorService.execute(() -> resultProcessor.process(process));
+            executorService.execute(() -> processData(jobId, dataExtractor, process, resultProcessor));
+        });
+    }
+
+    private void processData(String jobId, DataFrameDataExtractor dataExtractor, AnalyticsProcess process,
+                             AnalyticsResultProcessor resultProcessor) {
+        try {
+            writeHeaderRecord(dataExtractor, process);
+            writeDataRows(dataExtractor, process);
+            process.writeEndOfDataMessage();
+            process.flushStream();
 
-                LOGGER.debug("[{}] Closing process", jobId);
+            LOGGER.info("[{}] Waiting for result processor to complete", jobId);
```

Comment:
I don't think we'll want this to be an `INFO` in production.  If you want to leave it like this on the feature branch please add a `TODO` to downgrade it.

Reply:
Indeed, this shouldn't be an info message. I left it on purpose as I think it's useful during development. I had in mind that we'd review all logging when this is refactored into persistent tasks. Not sure I'd have a todo for each one of them, but I can add them if you think it's best.